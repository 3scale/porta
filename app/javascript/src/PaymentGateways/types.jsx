// @flow

export type BraintreeFormProps = {
  // eslint-disable-next-line flowtype/no-weak-types
  braintreeClient: any,
  billingAddress?: string,
  threeDSecureEnabled: boolean,
  formActionPath: string,
  countriesList: string,
  selectedCountryCode?: string
}
