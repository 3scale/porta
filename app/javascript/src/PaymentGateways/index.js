// @flow

export { StripeFormWrapper } from './stripe/components/StripeFormWrapper'
export { StripeCardForm } from './stripe/components/StripeCardForm'
export { BraintreeForm } from './braintree/BraintreeForm'
export { BraintreeFormWrapper } from './braintree/BraintreeFormWrapper'
export { BraintreeCardFields } from './braintree/BraintreeCardFields'
export { BraintreeBillingAddressFields } from './braintree/BraintreeBillingAddressFields'
export { BraintreeUserFields } from './braintree/BraintreeUserFields'
export { BraintreeSubmitFields } from './braintree/BraintreeSubmitFields'
export * from './braintree/braintree'
export { Label } from './braintree/components/Label'
export { Input } from './braintree/components/Input'
export { ListItem } from './braintree/components/ListItem'
export * from './braintree/types'
