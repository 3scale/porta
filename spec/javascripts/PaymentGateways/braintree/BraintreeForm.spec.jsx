// @flow

import * as React from 'react'

import { BraintreeForm } from 'PaymentGateways'
import { mount } from 'enzyme'

jest.mock('braintree-web/hosted-fields')
import * as hostedFields from 'braintree-web/hosted-fields'
jest.spyOn(hostedFields, 'create').mockImplementation(() => Promise.resolve({
  getState: () => ({ fields: {} }),
  on: (event, fn) => {
    if (event === 'validityChange') {
      fn()
    }
  }
}))

jest.mock('validate.js')
import * as validate from 'validate.js'

const COUNTRIES_LIST = '[["Afghanistan","AF"],["Albania","AL"],["Algeria","DZ"],["Spain","ES"]]'

const props = {
  braintreeClient: {},
  billingAddress: {
    company: 'Kserol',
    address: '',
    address1: 'Napols 182',
    address2: '',
    phone_number: '1234567890',
    city: 'Barcelona',
    country: 'Spain',
    state: '',
    zip: '08013'
  },
  threeDSecureEnabled: true,
  formActionPath: 'form-path',
  countriesList: COUNTRIES_LIST,
  selectedCountryCode: 'ES'
}

it('should render properly', () => {
  const wrapper = mount(<BraintreeForm {...props} />)
  expect(wrapper.exists()).toBe(true)
})

it('should render submit button disabled by default', () => {
  const wrapper = mount(<BraintreeForm {...props} />)
  expect(wrapper.find('.btn-primary').prop('disabled')).toEqual(true)
})

it('should pre-fill billing address inputs when a value is provided', () => {
  const wrapper = mount(<BraintreeForm {...props} />)
  expect(wrapper.find('input#customer_credit_card_billing_address_company').props().value).toEqual('Kserol')
  expect(wrapper.find('input#customer_credit_card_billing_address_street_address').props().value).toEqual('Napols 182')
  expect(wrapper.find('input#customer_credit_card_billing_address_postal_code').props().value).toEqual('08013')
  expect(wrapper.find('input#customer_credit_card_billing_address_locality').props().value).toEqual('Barcelona')
  expect(wrapper.find('select#customer_credit_card_billing_address_country_name').props().value).toEqual('ES')
})

// $FlowIgnore supress console error until FIXME is resolved to keep tests output clean
console.error = () => {}

// FIXME: Fix log error 'Warning: An update to BraintreeForm inside a test was not wrapped in act'
// Using 'act' or other solutions like 'runAllImmediates' does not seem to be valid solutions. Instead, the component BraintreeForm
// probably has to be refactored to reduce the number of async effects. Upgrading Enzyme to 3.10 is also irrelevant.
// Refs:
// - https://stackoverflow.com/q/55388587/5466997
// - https://github.com/eps1lon/react-act-immediate/blob/4d61b67dc98dd8dd422a41d07b82e08a8031bded/src/index.test.js
// - https://github.com/airbnb/enzyme/issues/2073
it('should enable submit button when form is valid', async () => {
  jest.spyOn(validate, 'validate')
    .mockReturnValueOnce(undefined)

  const wrapper = await mount(<BraintreeForm {...props} />)
  wrapper.update()
  expect(wrapper.find('.btn-primary').prop('disabled')).toEqual(false)
})
