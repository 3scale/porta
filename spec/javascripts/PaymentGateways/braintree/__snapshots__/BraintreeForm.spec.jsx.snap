// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<BraintreeForm
  billingAddress={
    Object {
      "address": "",
      "address1": "Napols 182",
      "address2": "",
      "city": "Barcelona",
      "company": "Kserol",
      "country": "Spain",
      "phone_number": "1234567890",
      "state": "",
      "zip": "08013",
    }
  }
  braintreeClient={Object {}}
  countriesList="[[\\"Afghanistan\\",\\"AF\\"],[\\"Albania\\",\\"AL\\"],[\\"Algeria\\",\\"DZ\\"],[\\"Spain\\",\\"ES\\"]]"
  formActionPath="form-path"
  selectedCountryCode="ES"
  threeDSecureEnabled={true}
>
  <form
    action="form-path"
    className="form-horizontal customer"
    id="customer_form"
    onChange={[Function]}
  >
    <input
      name="utf8"
      type="hidden"
      value="âœ“"
    />
    <CSRFToken>
      <input
        type="hidden"
      />
    </CSRFToken>
    <fieldset>
      <p
        className="required-fields"
      >
        All fields marked with an asterisk ( * ) are mandatory
      </p>
      <BraintreeUserFields>
        <ul
          className="list-unstyled"
        >
          <ListItem
            id="customer_first_name_input"
          >
            <li
              className="string optional form-group"
              id="customer_first_name_input"
            >
              <Label
                htmlFor="customer_first_name"
                label="First name"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_first_name"
                >
                  First name *
                </label>
              </Label>
              <Input
                id="customer_first_name"
                name="customer[first_name]"
                onChange={[Function]}
                required={true}
                value=""
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_first_name"
                  name="customer[first_name]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_last_name_input"
          >
            <li
              className="string optional form-group"
              id="customer_last_name_input"
            >
              <Label
                htmlFor="customer_last_name"
                label="Last name"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_last_name"
                >
                  Last name *
                </label>
              </Label>
              <Input
                id="customer_last_name"
                name="customer[last_name]"
                onChange={[Function]}
                required={true}
                value=""
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_last_name"
                  name="customer[last_name]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_phone_input"
          >
            <li
              className="string optional form-group"
              id="customer_phone_input"
            >
              <Label
                htmlFor="customer_phone"
                label="Phone"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_phone"
                >
                  Phone *
                </label>
              </Label>
              <Input
                id="customer_phone"
                name="customer[phone]"
                onChange={[Function]}
                value=""
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_phone"
                  name="customer[phone]"
                  onChange={[Function]}
                  type="text"
                  value=""
                />
              </Input>
            </li>
          </ListItem>
        </ul>
      </BraintreeUserFields>
    </fieldset>
    <fieldset>
      <BraintreeCardFields>
        <legend>
          Credit Card
        </legend>
        <ul
          className="list-unstyled"
        >
          <ListItem
            id="customer_credit_card_number_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_number_input"
            >
              <Label
                htmlFor="customer_credit_card_number"
                label="Number"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_number"
                >
                  Number *
                </label>
              </Label>
              <div
                className="form-control col-md-6"
                data-name="customer[credit_card][number]"
                id="customer_credit_card_number"
              />
              <div
                className="col-md-6 col-md-offset-4 inline-hints"
              >
                Incorrect card number. Specify a valid credit card number
              </div>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_cvv_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_cvv_input"
            >
              <Label
                htmlFor="customer_credit_card_cvv"
                label="CVV"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_cvv"
                >
                  CVV *
                </label>
              </Label>
              <div
                className="form-control col-md-6"
                data-name="customer[credit_card][cvv]"
                id="customer_credit_card_cvv"
              />
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_expiration_date_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_expiration_date_input"
            >
              <Label
                htmlFor="customer_credit_card_expiration_date"
                label="Expiration Date (MM/YY)"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_expiration_date"
                >
                  Expiration Date (MM/YY) *
                </label>
              </Label>
              <div
                className="form-control col-md-6"
                data-name="customer[credit_card][expiration_date]"
                id="customer_credit_card_expiration_date"
              />
            </li>
          </ListItem>
        </ul>
      </BraintreeCardFields>
    </fieldset>
    <fieldset>
      <BraintreeBillingAddressFields
        billingAddressData={
          Object {
            "address": "",
            "address1": "Napols 182",
            "address2": "",
            "city": "Barcelona",
            "company": "Kserol",
            "country": "Spain",
            "phone_number": "1234567890",
            "state": "",
            "zip": "08013",
          }
        }
        countriesList={
          Array [
            Array [
              "Afghanistan",
              "AF",
            ],
            Array [
              "Albania",
              "AL",
            ],
            Array [
              "Algeria",
              "DZ",
            ],
            Array [
              "Spain",
              "ES",
            ],
          ]
        }
        selectedCountryCode="ES"
        setBillingAddressData={[Function]}
      >
        <legend>
          Biling address
        </legend>
        <ul
          className="list-unstyled"
        >
          <ListItem
            id="customer_credit_card_billing_address_company_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_company_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_company"
                label="Company"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_company"
                >
                  Company *
                </label>
              </Label>
              <Input
                id="customer_credit_card_billing_address_company"
                name="customer[credit_card][billing_address][company]"
                onChange={[Function]}
                required={true}
                value="Kserol"
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_credit_card_billing_address_company"
                  name="customer[credit_card][billing_address][company]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="Kserol"
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_billing_address_street_address_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_street_address_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_street_address"
                label="Street address"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_street_address"
                >
                  Street address *
                </label>
              </Label>
              <Input
                id="customer_credit_card_billing_address_street_address"
                name="customer[credit_card][billing_address][street_address]"
                onChange={[Function]}
                required={true}
                value="Napols 182"
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_credit_card_billing_address_street_address"
                  name="customer[credit_card][billing_address][street_address]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="Napols 182"
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_billing_address_postal_code_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_postal_code_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_postal_code"
                label="ZIP / Postal Code"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_postal_code"
                >
                  ZIP / Postal Code *
                </label>
              </Label>
              <Input
                id="customer_credit_card_billing_address_postal_code"
                name="customer[credit_card][billing_address][postal_code]"
                onChange={[Function]}
                required={true}
                value="08013"
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_credit_card_billing_address_postal_code"
                  name="customer[credit_card][billing_address][postal_code]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="08013"
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_billing_address_locality_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_locality_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_locality"
                label="City"
                required={true}
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_locality"
                >
                  City *
                </label>
              </Label>
              <Input
                id="customer_credit_card_billing_address_locality"
                name="customer[credit_card][billing_address][locality]"
                onChange={[Function]}
                required={true}
                value="Barcelona"
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_credit_card_billing_address_locality"
                  name="customer[credit_card][billing_address][locality]"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="Barcelona"
                />
              </Input>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_billing_address_region_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_region_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_region"
                label="State/Region"
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_region"
                >
                  State/Region 
                </label>
              </Label>
              <Input
                id="customer_credit_card_billing_address_region"
                maxLength="3"
                name="customer[credit_card][billing_address][region]"
                onChange={[Function]}
                value=""
              >
                <input
                  className="col-md-6 form-control"
                  id="customer_credit_card_billing_address_region"
                  maxLength="3"
                  name="customer[credit_card][billing_address][region]"
                  onChange={[Function]}
                  type="text"
                  value=""
                />
              </Input>
              <div
                className="col-md-6 col-md-offset-4"
              >
                The 2 letter code for US states or an ISO-3166-2 country subdivision code of up to three letters. 
                <strong>
                  If unsure, leave blank
                </strong>
              </div>
            </li>
          </ListItem>
          <ListItem
            id="customer_credit_card_billing_address_country_name_input"
          >
            <li
              className="string optional form-group"
              id="customer_credit_card_billing_address_country_name_input"
            >
              <Label
                htmlFor="customer_credit_card_billing_address_country_name"
                label="Country"
              >
                <label
                  className="col-md-4 control-label"
                  htmlFor="customer_credit_card_billing_address_country_name"
                >
                  Country 
                </label>
              </Label>
              <select
                className="form-control col-md-6"
                id="customer_credit_card_billing_address_country_name"
                name="customer[credit_card][billing_address][country_name]"
                onChange={[Function]}
                required={true}
                value="ES"
              >
                <option
                  value=""
                />
                <option
                  key="AF0"
                  value="AF"
                >
                  Afghanistan
                </option>
                <option
                  key="AL1"
                  value="AL"
                >
                  Albania
                </option>
                <option
                  key="DZ2"
                  value="DZ"
                >
                  Algeria
                </option>
                <option
                  key="ES3"
                  value="ES"
                >
                  Spain
                </option>
              </select>
            </li>
          </ListItem>
        </ul>
      </BraintreeBillingAddressFields>
    </fieldset>
    <fieldset>
      <BraintreeSubmitFields
        isFormValid={false}
        onSubmitForm={[Function]}
      >
        <div
          className="form-group"
        >
          <div
            className="col-md-10 operations"
          >
            <button
              className="btn btn-primary pull-right"
              disabled={true}
              onClick={[Function]}
            >
              Save details
            </button>
          </div>
        </div>
      </BraintreeSubmitFields>
    </fieldset>
    <input
      id="braintree_nonce"
      name="braintree[nonce]"
      type="hidden"
      value=""
    />
  </form>
</BraintreeForm>
`;
